### ISSUE TITLE
Intermittent WebApi Startup.cs failure - "An existing connection was forcibly closed by the remote host"

### ISSUE THREAD

**Issue by jborn** [IdentityServer/IdentityServer3#3071]
- [x] I read and understood how to enable [logging](https://identityserver.github.io/Documentation/docsv2/configuration/logging.html)
### Question / Issue

I'm using IdentityServer3 with the Client Credential Flow to Authorize a WebApi Rest request.  The WebApi Rest request randomly throws an exception during startup complaining about "IOException: Unable to get document from: https://www.example.com/core/.well-known/openid-configuration]" and "InvalidOperationException: IDX10803: Unable to create to obtain configuration from: 'https://www.example.com/core/.well-known/openid-configuration'.]"  I don't see any relevant logs in the WebApi trace level logs either.

I setup the call to get the token and then call the Rest service using SoapUI on my local machine calling IdentityServer3 for the token on one QA server and then calling the WebApi on another QA server.  It works about 50% of the time.  When it stops working it can be a day before we get another request to work.

Any ideas on what's wrong?
### Stracktrace from SoapUI

```
[SocketException (0x2746): An existing connection was forcibly closed by the remote host]
   System.Net.Sockets.Socket.EndReceive(IAsyncResult asyncResult) +8156963
   System.Net.Sockets.NetworkStream.EndRead(IAsyncResult asyncResult) +48

[IOException: Unable to read data from the transport connection: An existing connection was forcibly closed by the remote host.]
   System.Net.Security._SslStream.EndRead(IAsyncResult asyncResult) +8111720
   System.Net.TlsStream.EndRead(IAsyncResult asyncResult) +275
   System.Net.Connection.ReadCallback(IAsyncResult asyncResult) +45

[WebException: The underlying connection was closed: An unexpected error occurred on a receive.]
   System.Net.HttpWebRequest.EndGetResponse(IAsyncResult asyncResult) +764
   System.Net.Http.HttpClientHandler.GetResponseCallback(IAsyncResult ar) +78

[HttpRequestException: An error occurred while sending the request.]

[AggregateException: One or more errors occurred.]
   System.Threading.Tasks.Task`1.GetResultCore(Boolean waitCompletionNotification) +4451240
   Microsoft.IdentityModel.Protocols.<GetDocumentAsync>d__0.MoveNext() in c:\workspace\WilsonForDotNet45Release\src\Microsoft.IdentityModel.Protocol.Extensions\Configuration\HttpDocumentRetriever.cs:53

[IOException: Unable to get document from: https://www.example.com/core/.well-known/openid-configuration]
   Microsoft.IdentityModel.Protocols.<GetDocumentAsync>d__0.MoveNext() in c:\workspace\WilsonForDotNet45Release\src\Microsoft.IdentityModel.Protocol.Extensions\Configuration\HttpDocumentRetriever.cs:59
   System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) +13847892
   System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) +61
   Microsoft.IdentityModel.Protocols.<GetAsync>d__0.MoveNext() in c:\workspace\WilsonForDotNet45Release\src\Microsoft.IdentityModel.Protocol.Extensions\Configuration\OpenIdConnectConfigurationRetriever.cs:81
   System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) +13847892
   System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) +61
   Microsoft.IdentityModel.Protocols.<GetConfigurationAsync>d__3.MoveNext() in c:\workspace\WilsonForDotNet45Release\src\Microsoft.IdentityModel.Protocol.Extensions\Configuration\ConfigurationManager.cs:0

[InvalidOperationException: IDX10803: Unable to create to obtain configuration from: 'https://www.example.com/core/.well-known/openid-configuration'.]
   Microsoft.IdentityModel.Protocols.<GetConfigurationAsync>d__3.MoveNext() in c:\workspace\WilsonForDotNet45Release\src\Microsoft.IdentityModel.Protocol.Extensions\Configuration\ConfigurationManager.cs:212
   System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) +13847892
   System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) +61
   Microsoft.IdentityModel.Protocols.<GetConfigurationAsync>d__0.MoveNext() in c:\workspace\WilsonForDotNet45Release\src\Microsoft.IdentityModel.Protocol.Extensions\Configuration\ConfigurationManager.cs:0
   System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) +13847892
   System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) +61
   System.Runtime.CompilerServices.TaskAwaiter`1.GetResult() +31
   IdentityServer.WebApi.AccessTokenValidation.<<RetrieveMetadata>b__0>d__4.MoveNext() in e:\Source Code\GitHub\IdentityServer\IdentityServer.WebApi.AccessTokenValidation\Plumbing\DiscoveryDocumentIssuerSecurityTokenProvider.cs:123
   System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) +13847892
   System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) +61
   System.Runtime.CompilerServices.TaskAwaiter`1.GetResult() +31
   IdentityServer.WebApi.AccessTokenValidation.AsyncHelper.RunSync(Func`1 func) in e:\Source Code\GitHub\IdentityServer\IdentityServer.WebApi.AccessTokenValidation\Plumbing\AsyncHelper.cs:18
   IdentityServer.WebApi.AccessTokenValidation.DiscoveryDocumentIssuerSecurityTokenProvider.RetrieveMetadata() in e:\Source Code\GitHub\IdentityServer\IdentityServer.WebApi.AccessTokenValidation\Plumbing\DiscoveryDocumentIssuerSecurityTokenProvider.cs:141
   IdentityServer.WebApi.AccessTokenValidation.DiscoveryDocumentIssuerSecurityTokenProvider..ctor(String discoveryEndpoint, IdentityServerBearerTokenAuthenticationOptions options, ILoggerFactory loggerFactory) in e:\Source Code\GitHub\IdentityServer\IdentityServer.WebApi.AccessTokenValidation\Plumbing\DiscoveryDocumentIssuerSecurityTokenProvider.cs:43
   Owin.IdentityServerBearerTokenValidationAppBuilderExtensions.ConfigureLocalValidation(IdentityServerBearerTokenAuthenticationOptions options, ILoggerFactory loggerFactory) in e:\Source Code\GitHub\IdentityServer\IdentityServer.WebApi.AccessTokenValidation\IdentityServerBearerTokenValidationAppBuilderExtensions.cs:129
   Owin.IdentityServerBearerTokenValidationAppBuilderExtensions.UseIdentityServerBearerTokenAuthentication(IAppBuilder app, IdentityServerBearerTokenAuthenticationOptions options) in e:\Source Code\GitHub\IdentityServer\IdentityServer.WebApi.AccessTokenValidation\IdentityServerBearerTokenValidationAppBuilderExtensions.cs:39
   Company.WebApi.waAddressQualification.Startup.Configuration(IAppBuilder app) in e:\Source Code\GitHub\waDemo\Startup.cs:23

[TargetInvocationException: Exception has been thrown by the target of an invocation.]
   System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor) +0
   System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(Object obj, Object[] parameters, Object[] arguments) +128
   System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture) +146
   Owin.Loader.<>c__DisplayClass12.<MakeDelegate>b__b(IAppBuilder builder) +93
   Owin.Loader.<>c__DisplayClass1.<LoadImplementation>b__0(IAppBuilder builder) +209
   Microsoft.Owin.Host.SystemWeb.OwinAppContext.Initialize(Action`1 startup) +843
   Microsoft.Owin.Host.SystemWeb.OwinBuilder.Build(Action`1 startup) +51
   Microsoft.Owin.Host.SystemWeb.OwinHttpModule.InitializeBlueprint() +101
   System.Threading.LazyInitializer.EnsureInitializedCore(T& target, Boolean& initialized, Object& syncLock, Func`1 valueFactory) +137
   Microsoft.Owin.Host.SystemWeb.OwinHttpModule.Init(HttpApplication context) +172
   System.Web.HttpApplication.RegisterEventSubscriptionsWithIIS(IntPtr appContext, HttpContext context, MethodInfo[] handlers) +618
   System.Web.HttpApplication.InitSpecial(HttpApplicationState state, MethodInfo[] handlers, IntPtr appContext, HttpContext context) +172
   System.Web.HttpApplicationFactory.GetSpecialApplicationInstance(IntPtr appContext, HttpContext context) +402
   System.Web.Hosting.PipelineRuntime.InitializeApplication(IntPtr appContext) +343

[HttpException (0x80004005): Exception has been thrown by the target of an invocation.]
   System.Web.HttpRuntime.FirstRequestInit(HttpContext context) +579
   System.Web.HttpRuntime.EnsureFirstRequestInit(HttpContext context) +112
   System.Web.HttpRuntime.ProcessRequestNotificationPrivate(IIS7WorkerRequest wr, HttpContext context) +712

Version Information: Microsoft .NET Framework Version:4.0.30319; ASP.NET Version:4.6.1055.0
```

**Comment by leastprivilege** [comment]
Never seen that before - do you have any "unusual" network devices between the API and the discovery endpoint?

**Comment by pianzide1117** [comment]
I got it too. I put identityserver and UseIdentityServerBearerTokenAuthentication in one webapi site , and still suffering.

**Comment by pianzide1117** [comment]
I get a solution for my situation

`
app.UseIdentityServer(options);
app.Map("/api",apiapp=>{
apiapp.UseIdentityServerBearerTokenAuthentication(new
IdentityServerBearerTokenAuthenticationOptions
{
Authority = WebConfigurationManager.AppSettings["Authority"],
RequiredScopes = new[] { "read", "write" }
});
})

`

**Comment by leastprivilege** [comment]
@pianzide1117  if IS and API are in the same application, you need to enable delay load on the validation middleware. that's unrelated though.

**Comment by pianzide1117** [comment]
@leastprivilege many thanks

**Comment by jborn** [comment]
@leastprivilege I wouldn't think the DelayLoadMetadata flag would help in my situation.  IS3 is running on a different server under a different IIS (Has to be with a different server) Not sure what we are using in our QA environment for the load balancer appliance, I'm reaching out to the networking team and will update when I get information back.

I believe our installation is standard however.  Both servers are Windows Server 2008 R2 Standard (Both are also VMs) Both are running IIS 7.5

**Comment by jborn** [comment]
Meant to add the contents of my Startup.cs also:
```

using Microsoft.Owin;
using Owin;
using System.Collections.Generic;
using System.IdentityModel.Tokens;
using System.Configuration;
using IdentityServer.WebApi.AccessTokenValidation;
using Midas.Core.Utilities.NLog;

[assembly: OwinStartup(typeof(DEMO.API.Startup))]

namespace TWC.MIDAS.WebApi.waAddressQualification
{
    public class Startup
    {
        private static readonly ICoreLog Logger = NLogLogger.GetNLogICoreLog("Startup");

        public void Configuration(IAppBuilder app)
        {
            Logger.Info("Application initializing.....");

            JwtSecurityTokenHandler.InboundClaimTypeMap = new Dictionary<string, string>();

            app.UseIdentityServerBearerTokenAuthentication(new IdentityServerBearerTokenAuthenticationOptions
                {
                    Authority = ConfigurationManager.AppSettings["IdentityServerUrl"],
                    RequiredScopes = new[] { ConfigurationManager.AppSettings["IdentityServerScopes"] }
                });

            app.UseWebApi(WebApiConfig.Register());

            Logger.Info("Application initialized.");
        }
    }
}
```

**Comment by jborn** [comment]
We are using an F5 load balancer with SSL offloading.

In the IdentityServerOptions.cs call we had to make the following changes to get the SSL offloading working:
```

            this.RequireSsl = false;
            this.PublicOrigin = System.Configuration.ConfigurationManager.AppSettings["IdentityServerPublicOrigin"]; 
```

IdentityServerPublicOrigin is the URL for our QA IdentityServer3 installation: "https://is.demo.com"

The error doesn't happen all the time, I can always hit the https://www.example.com/core/.well-known/openid-configuration URL

**Comment by brockallen** [comment]
All set on this issue?

**Comment by jborn** [comment]
We are still having this issue, but have been able to rule out a few things by using alternate URLs that are not behind the F5 load balancer.  To setup this test case I copied the API code to the new URL and changed the Identity Server URL to reflect the new setup.  The Identity Server code base used under the F5 was also copied to a new server that is not behind an F5.  Configuration was updated to remove the IdentityServerPublicOrigin setting under the appSettings in the Web.Config.  This setup works.

Another piece of the puzzle is while using SoapUI to get the token from the site behind the URL I'll often get a "Wed Aug 17 15:24:48 EDT 2016:ERROR:An error occurred [org.apache.oltu.oauth2.common.exception.OAuthSystemException: java.net.SocketException: Connection reset], see error log for details error message." in the SoapUI log.  This error message is very similar to the error being reported by the API and believe the entire issue is with the IdentityServer setup.  This could be the F5 is misconfigured or I've misconfigured the IdentityServer3 configuration.  The following is the IdentityServer3 configuration.  Is this correct for behind a load balancer?

        public IdentityServerOptions()
        {
            SiteName = Constants.IdentityServerName;

            this.ProtocolLogoutUrls = new List<string>();
            this.RequireSsl = false;
            this.PublicOrigin = System.Configuration.ConfigurationManager.AppSettings["IdentityServerPublicOrigin"];
            this.Endpoints = new EndpointOptions();
            this.AuthenticationOptions = new AuthenticationOptions();
            this.CspOptions = new CspOptions();
            this.LoggingOptions = new LoggingOptions();
            this.EventsOptions = new EventsOptions();
            this.EnableWelcomePage = true;
            this.InputLengthRestrictions = new InputLengthRestrictions();
        }

Any suggestions?

**Comment by jborn** [comment]
The URL in front of the the load balancer is https, but traffic between the load balancer and the QA site is http which is why I set the RequireSsl to false.  Is that correct?  Do I need to install certificates anywhere?

**Comment by brockallen** [comment]
Correct, if the traffic from LB to IdSvr is not SSL, then you need that flag. As for certs -- SSL certs will be on your LB. Signing certs will be on IdSvr.

**Comment by jborn** [comment]
And if it works intermittently then the certs are installed in the correct spot?  I think all I'm left with is a load balancer that is miss configured.

**Comment by brockallen** [comment]
Perhaps. It's hard to tell from where I am :/

**Comment by brockallen** [comment]
Any update?

**Comment by jborn** [comment]
I believe a configuration has changed, because I'm now getting a different error.  I have RequireSsl as false, but I'm getting the following on every access right now:

<data contentType="null" contentLength="15"><![CDATA[SSL is required]]></data>

This is the response from a SoapUI request to get the token.  It works in other environments, so why would it choke here?  Could a certificate be expired?

I've tried with RequireSsl as both false and true since I'm pulling them from my Web.Config the test was easy to setup.  Both give SSL is required above.  Tried running code locally and it worked just fine.

Here is the code that sets RequireSsl to false:
       public IdentityServerOptions()
        {
            SiteName = Constants.IdentityServerName;

            this.ProtocolLogoutUrls = new List<string>();
            bool requireSsl = false;
            Boolean.TryParse(System.Configuration.ConfigurationManager.AppSettings["RequireSSL"], out requireSsl);
            this.RequireSsl = requireSsl;
            this.PublicOrigin = System.Configuration.ConfigurationManager.AppSettings["IdentityServerPublicOrigin"];
            this.Endpoints = new EndpointOptions();
            this.AuthenticationOptions = new AuthenticationOptions();
            this.CspOptions = new CspOptions();
            this.LoggingOptions = new LoggingOptions();
            this.EventsOptions = new EventsOptions();
            this.EnableWelcomePage = true;
            this.InputLengthRestrictions = new InputLengthRestrictions();
        }

**Comment by brockallen** [comment]
Rereading your issue makes me think it might be something related to this: https://github.com/IdentityServer/IdentityServer3/issues/3253

As for your SSL issue, no idea.

**Comment by brockallen** [comment]
Can we close this issue?

**Comment by jborn** [comment]
Finally found the solution.  I have no idea why it was working some of the time, but it was.  I suspected that the issue was something to due with a misconfigured load balancer, but could never prove it and cannot personally see the setup of the load balancer.  I'm not a network guy, so this is probably a good thing.  Anyway it has something to do with a SNAT Pool.  I think the load balancer referenced the box name directly, but it should have been referencing the SNAT Pool instead.  Once updated it worked correctly.  I'm not sure if IdentityServer3 would be able to validate that type of misconfiguration, or if there is a tool to help determine that?  The frustrating part was it worked 10% of the time, then would not work for 90%.  Once the configuration was updated it works 100% of the time!
